### PR TITLE
Fix: Ignore template literals in no-mixed-tabs-and-spaces (fixes #4054)

### LINKED ISSUE #4054: Rule `no-mixed-spaces-and-tabs` should not apply inside template literals
```
~/dev
❯ eslint --version                                                               
v1.6.0
                                                                                                                                               
~/dev
❯ echo '`\n\t  `' | eslint --stdin --env=es6 --rule='no-mixed-spaces-and-tabs: 2'

<text>
  2:2  error  Mixed spaces and tabs  no-mixed-spaces-and-tabs

✖ 1 problem (1 error, 0 warnings)
```

This rule should not apply inside a template literal. I should be able to put whatever I want inside it.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/27180663-rule-no-mixed-spaces-and-tabs-should-not-apply-inside-template-literals?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by nzakas**

### CODE REVIEW

File: lib/rules/no-mixed-spaces-and-tabs.js [hunk 1]
```diff
@@ -26,13 +29,57 @@ module.exports = function(context) {
...
 
+    /**
+     * Determines if a given line and column are contained within a location.
+     * @param {Location} loc The location object from an AST node.
+     * @param {int} line The line to check.
+     * @param {int} column The column to check.
+     * @returns {boolean} True if the line and column are inside the location, false if not.
+     * @private
+     */
+    function inLoc(loc, line, column) {
```
Comment (xjamundx) [inline lib/rules/no-mixed-spaces-and-tabs.js]: Good candidate for a utility file?

File: lib/rules/no-mixed-spaces-and-tabs.js [hunk 2]
```diff
@@ -54,9 +101,19 @@ module.exports = function(context) {
                 match = regex.exec(line);
 
                 if (match) {
+                    var lineNumber = i + 1,
+                        column = match.index + 1;
+
+                    // make sure this isn't inside of a template element
+                    for (; lastTemplateLocIndex < templateLocs.length; lastTemplateLocIndex++) {
+
+                        if (inLoc(templateLocs[lastTemplateLocIndex], lineNumber, column)) {
```
Comment (xjamundx) [inline lib/rules/no-mixed-spaces-and-tabs.js]: What is you have a template literal and a mixed tabs and spacing on the same line? is that too obtuse. Is there a test for that case?

File: lib/rules/no-mixed-spaces-and-tabs.js (hunk 1 above)
Comment (nzakas) [inline lib/rules/no-mixed-spaces-and-tabs.js]: Thought about that, but didn't see an obvious other use for it. Will keep an eye out going forward.

File: lib/rules/no-mixed-spaces-and-tabs.js (hunk 2 above)
Comment (nzakas) [inline lib/rules/no-mixed-spaces-and-tabs.js]: Yup, there's a test for that. It's covered.

**Comment by ilyavolodin** [comment]
LGTM

**Pull request merged**